libvast/src/option_declaration_set.cpp

```diff
@@ -0,0 +1,296 @@
+  for (auto& [_, x] : long_opts_) {
+    res << "\n"
+        << std::left << std::setw(help_column_width) << build_argument(*x)
+        << " : " << x->description();
+  }
+  return res.str();
+}
+
+size_t option_declaration_set::size() const {
+  return long_opts_.size();
+}
+
+expected<void> option_declaration_set::add(std::string_view name,
+                                           std::string_view desciption,
+                                           data default_value) {
+  // Parse short and long name.
+  std::string_view long_name;
+  std::vector<char> short_names;
+  if (auto idx = name.find(','); idx == std::string_view::npos) {
+    long_name = name;
+  } else {
+    long_name = name.substr(0, idx);
+    auto short_name = name.substr(idx + 1, name.size() - idx);
+    short_names.reserve(short_name.size());
+    for (auto x : short_name)
+      short_names.emplace_back(x);
+  }
+  // Validate short and long name.
+  if (long_name.empty())
+    return make_error(ec::unspecified, "no long-name specified");
+  // TODO: Remove explicit conversion to a string
+  if (auto it = long_opts_.find(std::string{long_name}); it != long_opts_.end())
+    return make_error(ec::unspecified, "long-name: " + std::string{long_name}
+                                         + " already in use");
+  for (auto x : short_names)
+    if (auto it = short_opts_.find(x); it != short_opts_.end())
+      return make_error(ec::unspecified,
+                        "short-name: " + to_string(x) + " already in use");
+  // Update option_declaration_set.
+  auto has_argument = visit(
+    [](const auto& arg) {
+      using arg_type = typename std::decay<decltype(arg)>::type;
+      return !std::is_same<arg_type, bool>::value;
+    }, default_value);
+  auto option = std::make_shared<option_declaration>(
+    long_name, std::move(short_names), desciption, has_argument,
+    std::move(default_value));
+  long_opts_.insert(std::make_pair(option->long_name(), option));
+  for (auto x : option->short_names())
+    short_opts_.insert(std::make_pair(x, option));
+  return no_error;
+}
+
+std::pair<option_declaration_set::parse_state,
+          option_declaration_set::argument_iterator>
+option_declaration_set::parse(option_map& xs, argument_iterator begin,
+                              argument_iterator end) const {
+  // Add all default values to the map.
+  for (auto& [long_name, x] : long_opts_) {
+    if (auto res = xs.add(long_name, x->default_value()); !res)
+      if (long_name != "help")
+        return std::make_pair(parse_state::option_already_exists, end);
+  }
+  auto parse_argument = [&](auto idx, const auto& option, auto begin, auto end) {
+    auto make_result = [](auto state, auto it, auto result) {
+      return std::make_pair(std::make_pair(state, it), std::move(result));
+    };
+    if (begin == end)
+      return make_result(parse_state::arg_declared_but_not_passed, begin,
+                         data{});
+    auto arg = begin->substr(idx);
```

Comment:
`*begin` returns a `std::string`, doesn't it? In this case `substr` copies part of the string. We could avoid copying via `std::string_view{*begin}.substr(idx)`.